### PR TITLE
BBA Changes

### PR DESCRIPTION
**Pull request by T-Nicholls**
Changes made while attempting to get BBA to interface with the virtac.

### CODE REVIEW

**Comment by T-Nicholls** [comment]
The Python 3.7 build is failing due to `pytest-flake8` which is abandoned, it's not worth fixing it on its own as it'd be resolved whenever we move ATIP over to the Diamond pip-skeleton/ copier template/new deployment system.

File: virtac/atip_ioc_entry.py [hunk 1]
```diff
@@ -56,14 +60,39 @@ def main():
         not args.disable_emittance,
     )
 
-    # Add special case out record for SOFB to write to.
-    builder.SetDeviceName("CS-CS-MSTAT-01")
-    builder.aOut("FBHEART", initial_value=10)
+    # Warn if set to default EPICS port(s) as this will likely casue PV conflicts.
```
Comment (MJGaughran) [inline virtac/atip_ioc_entry.py:63]: cause

File: atip/simulator.py [hunk 2]
```diff
@@ -186,35 +198,55 @@ def _recalculate_phys_data(self, callback):
...
+                # Signal the up to date flag since the physics data is now up to date.
+                # We do this before the callback is executed in case the callback
+                # checks the flag.
                 self.up_to_date.Signal()
                 if callback is not None:
                     logging.debug("Executing callback function.")
                     callback()
                     logging.debug("Callback completed.")
+            # This could cause thread-locking issues if another application using
+            # cothreads is poorly written and running on the same machine, but this
```
Comment (MJGaughran) [inline atip/simulator.py:210]: So a completely separate application running on the same machine can cause problems here?

Or do you mean on the same Python process?

File: virtac/atip_ioc_entry.py [hunk 3]
```diff
@@ -56,14 +60,39 @@ def main():
...
+    conflict_warning = ", this may lead to conflicting PV names on multiple servers."
```
Comment (MJGaughran) [inline virtac/atip_ioc_entry.py:64]: "this may lead to conflicting PV names with production IOCs"?

File: virtac/atip_ioc_entry.py [hunk 4]
```diff
@@ -21,14 +23,16 @@ def parse_arguments():
     parser.add_argument(
         "--disable-emittance", "-d", help="disable emittance calc", action="store_true"
     )
+    parser.add_argument(
+        "--enable-tfb", "-t", help="enable dummy Tune Feedback", action="store_true"
```
Comment (MJGaughran) [inline virtac/atip_ioc_entry.py:27]: Should this be 'enable dummy Tune Feedback PVs'?

Comment (MJGaughran) [reply virtac/atip_ioc_entry.py:27]: Is this dummy PVs pretending to be the TFB system, or PVs pretending to be the underlying hardware required for TFB to operate?

It's still not clear from the description

File: atip/simulator.py (hunk 2 above)
Comment (T-Nicholls) [reply atip/simulator.py:210]: Yes, separate application, doesn't have to be the same Python process

File: virtac/atip_ioc_entry.py (hunk 4 above)
Comment (T-Nicholls) [reply virtac/atip_ioc_entry.py:27]: The underlying hardware required for TFB to operate

File: virtac/atip_ioc_entry.py (hunk 1 above)
Comment (T-Nicholls) [reply virtac/atip_ioc_entry.py:63]: fixed

File: virtac/atip_ioc_entry.py (hunk 3 above)
Comment (T-Nicholls) [reply virtac/atip_ioc_entry.py:64]: changed

File: virtac/atip_ioc_entry.py (hunk 4 above)
Comment (T-Nicholls) [reply virtac/atip_ioc_entry.py:27]: I've gone with "enable dummy hardware to be used by Tune Feedback", hopefully that's clearer

**Review by MJGaughran — APPROVED** [review]
If you're able to be more explicit about the cause of errors with other cothread applications, I would suggest adding that but otherwise this PR is good.

**Comment by T-Nicholls** [comment]
I don't think I can elaborate concisely on the issue with other cothread applications.

Side note, I noticed that the fix for the issue in AT that was blocking us from making I04 the default ring mode got merged, so I've updated it and snuck it on the end of this pr.

Also, for some reason, I don't have merge permissions for this repo or Pytac, but I do for dls-bba

**Pull request merged**